hw/opentitan/ot_usbdev.c

```diff
@@ -664,6 +1011,315 @@ static void ot_usbdev_simulate_link_reset(OtUsbdevState *s)
+        ot_usbdev_update_irqs(s);
+        return;
+    }
+
+    trace_ot_usbdev_setup(s->ot_id, ep, "accepted");
+
+    /*
+     * Obtain the next SETUP buffer ID and write the SETUP packet to the buffer
+     */
+    uint8_t buf_id = fifo8_pop(&s->av_setup_fifo);
+    ot_usbdev_write_buffer(s, buf_id, setup, OT_USBDEV_SETUP_PACKET_SIZE);
+
+    /* Create and push an entry in the RX FIFO */
+    uint32_t rx_fifo_entry = 0u;
+    rx_fifo_entry = FIELD_DP32(rx_fifo_entry, RXFIFO, BUFFER, buf_id);
+    rx_fifo_entry = FIELD_DP32(rx_fifo_entry, RXFIFO, SETUP, 1u);
+    rx_fifo_entry = FIELD_DP32(rx_fifo_entry, RXFIFO, EP, ep);
+    rx_fifo_entry =
+        FIELD_DP32(rx_fifo_entry, RXFIFO, SIZE, OT_USBDEV_SETUP_PACKET_SIZE);
+    fifo32_push(&s->rx_fifo, rx_fifo_entry);
+
+    /* @todo Cancel any pending transfer on EP */
+
+    /* clear any STALL condition */
+    s->regs[R_OUT_STALL] &= ~(1u << ep);
+    s->regs[R_IN_STALL] &= ~(1u << ep);
+
+    /* Cancel any pending IN packet on this endpoint */
+    ot_usbdev_retire_in_packets(s, ep);
+
+    ot_usbdev_update_fifos(s);
+    ot_usbdev_update_irqs(s);
+}
+
+/*
+ * TODO
+ */
+static void ot_usbdev_complete_transfer_in(OtUsbdevState *s, uint8_t epnum,
+                                           OtUsbdevServerTransferStatus status,
+                                           const char *msg)
+{
+    OtUsbdevServer *server = &s->usb_server;
+    OtUsbdevServerEpInXfer *xfer = &server->ep[epnum].in;
+
+    uint32_t xfered_len = xfer->xfer_len - xfer->recv_rem;
+    ot_usbdev_server_complete_transfer(s, xfer->id, status, xfered_len,
+                                       xfer->xfer_buf, xfered_len, msg);
+
+    /* cleanup */
+    g_free(xfer->xfer_buf);
+    memset(xfer, 0u, sizeof(OtUsbdevServerEpInXfer));
+}
+
+/*
+ * TODO
+ */
+static void ot_usbdev_complete_transfer_out(OtUsbdevState *s, uint8_t epnum,
+                                            OtUsbdevServerTransferStatus status,
+                                            const char *msg)
+{
+    OtUsbdevServer *server = &s->usb_server;
+    OtUsbdevServerEpOutXfer *xfer = &server->ep[epnum].out;
+
+    uint32_t xfered_len = xfer->xfer_len - xfer->send_rem;
+    ot_usbdev_server_complete_transfer(s, xfer->id, status, xfered_len, NULL,
+                                       0u, msg);
+
+    /* cleanup */
+    g_free(xfer->xfer_buf);
+    memset(xfer, 0u, sizeof(OtUsbdevServerEpOutXfer));
+}
+
+/*
+ * TODO
+ */
+static void ot_usbdev_advance_transfer_in(OtUsbdevState *s, uint8_t epnum)
+{
+    g_assert(epnum < USBDEV_PARAM_N_ENDPOINTS);
+    /* See BFM (in_packet and handshake_packet) */
+
+    /* Nothing to do if there is no pending transfer */
+    OtUsbdevServer *server = &s->usb_server;
+    OtUsbdevServerEpInXfer *xfer = &server->ep[epnum].in;
+    if (!xfer->pending) {
+        return;
+    }
+
+    /*
+     * If the endpoint is not enabled, the device will silently ignored the
```

Comment:
```suggestion
     * If the endpoint is not enabled, the device will silently ignore the
```
(repeated in `ot_usbdev_advance_transfer_out`)